src/normalize/index.ts

```diff
@@ -1,28 +1,126 @@
 import {CoreNormalizer} from './core';
+import {deepEqual} from '../util';
 
-export function normalize(spec: TopLevelSpec, config?: Config): TopLevel<NormalizedSpec> {
+export function normalize(
+  spec: TopLevelSpec & LayoutSizeMixins,
+  config?: Config
+): TopLevel<NormalizedSpec> & LayoutSizeMixins {
   if (config === undefined) {
     config = initConfig(spec.config);
   }
 
-  return normalizeGenericSpec(spec, config);
+  let normalizedSpec = normalizeGenericSpec(spec, config);
+
+  normalizedSpec = normalizeAutoSize(
+    normalizedSpec,
+    {width: spec.width, height: spec.height, autosize: spec.autosize},
+    config
+  );
+
+  return normalizedSpec;
 }
 
 const normalizer = new CoreNormalizer();
 
 /**
  * Decompose extended unit specs into composition of pure unit specs.
  */
-function normalizeGenericSpec(
-  spec: GenericSpec<UnitSpec, ExtendedLayerSpec> | FacetedUnitSpec,
-  config: Config = {}
-): NormalizedSpec {
+function normalizeGenericSpec(spec: GenericSpec<UnitSpec, ExtendedLayerSpec> | FacetedUnitSpec, config: Config = {}) {
   return normalizer.map(spec, {config});
 }
 
+function _normalizeAutoSize(autosize: AutosizeType | AutoSizeParams) {
+  return isString(autosize) ? {type: autosize} : autosize || {};
+}
+
+/**
+ * Normalize autosize and deal with width or height == "container".
+ */
+export function normalizeAutoSize(
+  spec: TopLevel<NormalizedSpec>,
+  sizeInfo: {autosize: AutosizeType | AutoSizeParams} & LayoutSizeMixins,
+  config?: Config
+): TopLevel<NormalizedSpec> & LayoutSizeMixins {
+  let {width, height} = sizeInfo;
+
+  const isFitCompatible = isUnitSpec(spec) || isLayerSpec(spec);
+  const autosizeDefault: AutoSizeParams = {};
+
+  if (!isFitCompatible) {
+    // If spec is not compatible with autosize == "fit", discard width/height == container
+    if (width == 'container') {
+      log.warn(log.message.containerSizeNonSingle('width'));
+      width = undefined;
+    }
+    if (height == 'container') {
+      log.warn(log.message.containerSizeNonSingle('height'));
+      height = undefined;
+    }
+  } else {
+    // Default autosize parameters to fit when width/height is "container"
+    if (width == 'container' && height == 'container') {
+      autosizeDefault.type = 'fit';
+      autosizeDefault.contains = 'padding';
+    } else if (width == 'container') {
+      autosizeDefault.type = 'fit-x';
+      autosizeDefault.contains = 'padding';
+    } else if (height == 'container') {
+      autosizeDefault.type = 'fit-y';
+      autosizeDefault.contains = 'padding';
+    }
+  }
+
+  const autosize: AutoSizeParams = {
+    type: 'pad',
+    ...autosizeDefault,
+    ...(config ? _normalizeAutoSize(config.autosize) : {}),
+    ..._normalizeAutoSize(spec.autosize)
+  };
+
+  if (autosize.type === 'fit' && !isFitCompatible) {
+    log.warn(log.message.FIT_NON_SINGLE);
+    autosize.type = 'pad';
+  }
+
+  if (width == 'container' && !(autosize.type == 'fit' || autosize.type == 'fit-x')) {
```

Comment:
These warnings seem superfluous.  Using 'container' for the width and height seems to be the best way to be responsive to the container size.   And using autosize.type == 'none' is the only way (?) to specify the padding exactly, rather than adding any auto-computed padding based on what is drawn.  This is necessary to get predictable alignment across charts that are drawn separately.  So we would like to use this combination without getting log warnings.  Is that possible?

Reply:
Definitely. Wouldn't the issues be that the responsive size does not include the padding and therefore the chart is always too wide? If not, then I agree that we should remove this warning.

Reply:
Indeed, we would want contains: 'padding' to include the padding in the total.   So the content.width should be container.width - (padding.left + padding.right).  I don't really know if type: 'none' works this way, however.

```
 width: 'container',
 height: 'container',
 padding: { left: 40, right: 0, top: 5, bottom: 20 },
 autosize: { type: 'none', contains: 'padding' }
```

Reply:
It might not in Vega but we can generate the correct output. So if the user specifies autosize none, we can automatically change the container size computation to take the padding into account.

Happy to review a pull request towards this. I think the current behavior warrants a warning.